File: sample-chat-java/app/src/main/java/com/quickblox/sample/chat/java/ui/activity/LoginActivity.java
package com.quickblox.sample.chat.java.ui.activity;

import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.os.Vibrator;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.quickblox.core.QBEntityCallback;
import com.quickblox.core.exception.QBResponseException;
import com.quickblox.sample.chat.java.App;
import com.quickblox.sample.chat.java.R;
import com.quickblox.sample.chat.java.utils.SharedPrefsHelper;
import com.quickblox.sample.chat.java.utils.ValidationUtils;
import com.quickblox.sample.chat.java.utils.chat.ChatHelper;
import com.quickblox.users.QBUsers;
import com.quickblox.users.model.QBUser;

import java.util.Timer;
import java.util.TimerTask;

public class LoginActivity extends BaseActivity {
    private static final int UNAUTHORIZED = 401;
    private static final String DRAFT_LOGIN = "draft_login";
    private static final String DRAFT_USERNAME = "draft_username";

    private EditText loginEt;
    private EditText usernameEt;
    private TextView loginHint;
    private TextView usernameHint;
    private TextView btnLogin;
    private CheckBox chbSave;
    private LinearLayout rootView;
    private LinearLayout hidableHolder;

    public static void start(Context context) {
        Intent intent = new Intent(context, LoginActivity.class);
        context.startActivity(intent);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);
        initViews();
        prepareListeners();
        fillViews();
        defineFocusedBehavior();
    }

    private void initViews() {
        if (getSupportActionBar() != null) {
            getSupportActionBar().setElevation(20f);
        }
        loginHint = findViewById(R.id.tv_login_hint);
        usernameHint = findViewById(R.id.tv_username_hint);
        btnLogin = findViewById(R.id.tv_btn_login);
        loginEt = findViewById(R.id.et_login);
        usernameEt = findViewById(R.id.et_user_name);
        chbSave = findViewById(R.id.chb_login_save);
        btnLogin = findViewById(R.id.tv_btn_login);
        hidableHolder = findViewById(R.id.ll_hidable_holder);
        rootView = findViewById(R.id.root_view_login_activity);
    }

    private void fillViews() {
        String draftLogin = SharedPrefsHelper.getInstance().get(DRAFT_LOGIN, null);
        String draftUserName = SharedPrefsHelper.getInstance().get(DRAFT_USERNAME, null);

        if (!TextUtils.isEmpty(draftLogin)) {
            loginEt.setText(draftLogin);
        }
        if (!TextUtils.isEmpty(draftUserName)) {
            usernameEt.setText(draftUserName);
        }

        validateFields();
    }

    private void defineFocusedBehavior() {
        loginHint.setVisibility(View.GONE);
        usernameHint.setVisibility(View.GONE);

        loginEt.setOnFocusChangeListener(new View.OnFocusChangeListener() {

            @Override
            public void onFocusChange(View v, boolean hasFocus) {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                    if (hasFocus) {
                        loginEt.setTranslationZ(10f);
                    } else {
                        loginEt.setTranslationZ(0f);
                    }
                }
                if (ValidationUtils.isLoginValid(LoginActivity.this, loginEt)) {
                    loginHint.setVisibility(View.GONE);
                } else {
                    loginHint.setVisibility(View.VISIBLE);
                }
            }
        });

        usernameEt.setOnFocusChangeListener(new View.OnFocusChangeListener() {
            @Override
            public void onFocusChange(View v, boolean hasFocus) {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                    if (hasFocus) {
                        usernameEt.setTranslationZ(10f);
                    } else {
                        usernameEt.setTranslationZ(0f);
                    }
                }
                if (ValidationUtils.isLoginValid(LoginActivity.this, usernameEt)) {
                    usernameHint.setVisibility(View.GONE);
                } else {
                    usernameHint.setVisibility(View.VISIBLE);
                }
            }
        });
    }

    private void prepareListeners() {
        rootView.setOnLongClickListener(new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View v) {
                chbSave.setChecked(!chbSave.isChecked());

                Vibrator vibrator = (Vibrator) getSystemService(Context.VIBRATOR_SERVICE);
                if (chbSave.isChecked()) {
                    if (vibrator != null) {
                        vibrator.vibrate(80);
                    }
                } else {
                    if (vibrator != null) {
                        vibrator.vibrate(250);
                    }
                }
                return true;
            }
        });

        btnLogin.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (btnLogin.isActivated()) {
                    showProgressDialog(R.string.dlg_login);
                    prepareUser();
                }
            }
        });

        loginEt.addTextChangedListener(new TextWatcherListener(loginEt));
        usernameEt.addTextChangedListener(new TextWatcherListener(usernameEt));
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_activity_login, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_login_app_info:
                AppInfoActivity.start(this);
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    private Boolean validateFields() {
        Boolean loginValid = ValidationUtils.isLoginValid(this, loginEt);
        Boolean userNameValid = ValidationUtils.isFullNameValid(this, usernameEt);

        if (loginValid) {
            loginHint.setVisibility(View.GONE);
        } else {
            loginHint.setVisibility(View.VISIBLE);
        }

        if (userNameValid) {
            usernameHint.setVisibility(View.GONE);
        } else {
            usernameHint.setVisibility(View.VISIBLE);
        }

        if (loginValid && userNameValid) {
            btnLogin.setActivated(true);
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                btnLogin.setElevation(0F);
                btnLogin.setTranslationZ(10F);
            }
            return true;
        } else {
            btnLogin.setActivated(false);
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                btnLogin.setElevation(0F);
                btnLogin.setTranslationZ(0F);
            }
            return false;
        }
    }

    private void saveDrafts() {
        SharedPrefsHelper.getInstance().save(DRAFT_LOGIN, loginEt.getText().toString());
        SharedPrefsHelper.getInstance().save(DRAFT_USERNAME, usernameEt.getText().toString());
    }

    private void clearDrafts() {
        SharedPrefsHelper.getInstance().save(DRAFT_LOGIN, "");
        SharedPrefsHelper.getInstance().save(DRAFT_USERNAME, "");
    }

    private void prepareUser() {
        QBUser qbUser = new QBUser();
        qbUser.setLogin(loginEt.getText().toString().trim());
        qbUser.setFullName(usernameEt.getText().toString().trim());
        qbUser.setPassword(App.USER_DEFAULT_PASSWORD);
        signIn(qbUser);
    }

    private void signIn(final QBUser user) {
        showProgressDialog(R.string.dlg_login);
        ChatHelper.getInstance().login(user, new QBEntityCallback<QBUser>() {
            @Override
            public void onSuccess(QBUser userFromRest, Bundle bundle) {
                if (userFromRest.getFullName().equals(user.getFullName())) {
                    loginToChat(user);
                } else {
                    //Need to set password NULL, because server will update user only with NULL password
                    user.setPassword(null);
                    updateUser(user);
                }
            }

            @Override
            public void onError(QBResponseException e) {
                if (e.getHttpStatusCode() == UNAUTHORIZED) {
                    signUp(user);
                } else {
                    hideProgressDialog();
                    showErrorSnackbar(R.string.login_chat_login_error, e, new View.OnClickListener() {
                        @Override
                        public void onClick(View v) {
                            signIn(user);
                        }
                    });
                }
            }
        });
    }

    private void updateUser(final QBUser user) {
        ChatHelper.getInstance().updateUser(user, new QBEntityCallback<QBUser>() {
            @Override
            public void onSuccess(QBUser user, Bundle bundle) {
                loginToChat(user);
            }

            @Override
            public void onError(QBResponseException e) {
                hideProgressDialog();
                showErrorSnackbar(R.string.login_chat_login_error, e, null);
            }
        });
    }

    private void loginToChat(final QBUser user) {
        //Need to set password, because the server will not register to chat without password
        user.setPassword(App.USER_DEFAULT_PASSWORD);
        ChatHelper.getInstance().loginToChat(user, new QBEntityCallback<Void>() {
            @Override
            public void onSuccess(Void aVoid, Bundle bundle) {
                SharedPrefsHelper.getInstance().saveQbUser(user);
                if (!chbSave.isChecked()) {
                    clearDrafts();
                }
                DialogsActivity.start(LoginActivity.this);
                finish();
                hideProgressDialog();
            }

            @Override
            public void onError(QBResponseException e) {
                hideProgressDialog();
                showErrorSnackbar(R.string.login_chat_login_error, e, null);
            }
        });
    }

    private void signUp(final QBUser newUser) {
        SharedPrefsHelper.getInstance().removeQbUser();
        QBUsers.signUp(newUser).performAsync(new QBEntityCallback<QBUser>() {
            @Override
            public void onSuccess(QBUser user, Bundle bundle) {
                hideProgressDialog();
                signIn(newUser);
            }

            @Override
            public void onError(QBResponseException e) {
                hideProgressDialog();
                showErrorSnackbar(R.string.login_sign_up_error, e, null);
            }
        });
    }

    private class TextWatcherListener implements TextWatcher {
        private EditText editText;
        private Timer timer = new Timer();

        private TextWatcherListener(EditText editText) {
            this.editText = editText;
        }

        @Override
        public void beforeTextChanged(CharSequence s, int start, int count, int after) {

        }

        @Override
        public void onTextChanged(CharSequence s, int start, int before, int count) {
            String text = s.toString().replace("  ", " ");
            if (!editText.getText().toString().equals(text)) {
                editText.setText(text);
                editText.setSelection(text.length());
            }
            validateFields();
        }

        @Override
        public void afterTextChanged(Editable s) {
            timer.cancel();
            timer = new Timer();
            timer.schedule(new TimerTask() {
                @Override
                public void run() {
                    saveDrafts();
                }
            }, 300);
        }
    }
}

File: sample-chat-kotlin/app/src/main/java/com/quickblox/sample/chat/kotlin/ui/activity/LoginActivity.kt
package com.quickblox.sample.chat.kotlin.ui.activity

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.os.Vibrator
import android.text.Editable
import android.text.TextWatcher
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.CheckBox
import android.widget.EditText
import android.widget.LinearLayout
import android.widget.TextView
import com.quickblox.core.QBEntityCallback
import com.quickblox.core.exception.QBResponseException
import com.quickblox.sample.chat.kotlin.R
import com.quickblox.sample.chat.kotlin.USER_DEFAULT_PASSWORD
import com.quickblox.sample.chat.kotlin.utils.SharedPrefsHelper
import com.quickblox.sample.chat.kotlin.utils.chat.ChatHelper
import com.quickblox.sample.chat.kotlin.utils.isFullNameValid
import com.quickblox.sample.chat.kotlin.utils.isLoginValid
import com.quickblox.users.QBUsers
import com.quickblox.users.model.QBUser
import java.util.*


private const val UNAUTHORIZED = 401
private const val DRAFT_LOGIN = "draft_login"
private const val DRAFT_USERNAME = "draft_username"

class LoginActivity : BaseActivity() {

    private lateinit var loginEt: EditText
    private lateinit var usernameEt: EditText
    private lateinit var loginHint: TextView
    private lateinit var usernameHint: TextView
    private lateinit var btnLogin: TextView
    private lateinit var chbSave: CheckBox
    private lateinit var rootView: LinearLayout
    private lateinit var hidableHolder: LinearLayout

    companion object {
        fun start(context: Context) = context.startActivity(Intent(context, LoginActivity::class.java))
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login)
        initViews()
        prepareListeners()
        fillViews()
        defineFocusedBehavior()
    }

    private fun initViews() {
        supportActionBar?.elevation = 20f
        loginHint = findViewById(R.id.tv_login_hint)
        usernameHint = findViewById(R.id.tv_username_hint)
        btnLogin = findViewById(R.id.tv_btn_login)
        loginEt = findViewById(R.id.et_login)
        usernameEt = findViewById(R.id.et_user_name)
        chbSave = findViewById(R.id.chb_login_save)
        btnLogin = findViewById(R.id.tv_btn_login)
        hidableHolder = findViewById(R.id.ll_hidable_holder)
        rootView = findViewById(R.id.root_view_login_activity)
    }

    private fun fillViews() {
        if (!SharedPrefsHelper.get<String>(DRAFT_LOGIN).isNullOrEmpty()) {
            loginEt.setText(SharedPrefsHelper.get<String>(DRAFT_LOGIN))
        }
        if (!SharedPrefsHelper.get<String>(DRAFT_USERNAME).isNullOrEmpty()) {
            usernameEt.setText(SharedPrefsHelper.get<String>(DRAFT_USERNAME))
        }
        validateFields()
    }

    @SuppressLint("NewApi")
    private fun defineFocusedBehavior() {
        loginHint.visibility = View.GONE
        usernameHint.visibility = View.GONE

        loginEt.setOnFocusChangeListener { view, hasFocus ->
            if (hasFocus) {
                loginEt.translationZ = 10f
            } else {
                loginEt.translationZ = 0f
            }
            if (isLoginValid(this, loginEt)) {
                loginHint.visibility = View.GONE
            } else {
                loginHint.visibility = View.VISIBLE
            }
        }

        usernameEt.setOnFocusChangeListener { view, hasFocus ->
            if (hasFocus) {
                usernameEt.translationZ = 10f
            } else {
                usernameEt.translationZ = 0f
            }
            if (isFullNameValid(this, usernameEt)) {
                usernameHint.visibility = View.GONE
            } else {
                usernameHint.visibility = View.VISIBLE
            }
        }
    }

    private fun prepareListeners() {
        rootView.setOnLongClickListener(object : View.OnLongClickListener {
            override fun onLongClick(v: View?): Boolean {
                val vibrator = getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
                vibrator.vibrate(80)

                if (hidableHolder.visibility == View.GONE) {
                    hidableHolder.visibility = View.VISIBLE
                } else {
                    hidableHolder.visibility = View.GONE
                }
                return true
            }
        })

        btnLogin.setOnClickListener {
            if (btnLogin.isActivated) {
                showProgressDialog(R.string.dlg_login)
                prepareUser()
            }
        }

        loginEt.addTextChangedListener(TextWatcherListener(loginEt))
        usernameEt.addTextChangedListener(TextWatcherListener(usernameEt))
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.menu_activity_login, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.menu_login_app_info -> {
                AppInfoActivity.start(this)
                return true
            }

            else -> return super.onOptionsItemSelected(item)
        }
    }

    private fun validateFields(): Boolean {
        if (isLoginValid(this@LoginActivity, loginEt)) {
            loginHint.visibility = View.GONE
        } else {
            loginHint.visibility = View.VISIBLE
        }

        if (isFullNameValid(this@LoginActivity, usernameEt)) {
            usernameHint.visibility = View.GONE
        } else {
            usernameHint.visibility = View.VISIBLE
        }

        if (isLoginValid(this@LoginActivity, loginEt) && isFullNameValid(this@LoginActivity, usernameEt)) {
            btnLogin.isActivated = true
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                btnLogin.elevation = 0F
                btnLogin.translationZ = 10F
            }
            return true
        } else {
            btnLogin.isActivated = false
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                btnLogin.elevation = 0F
                btnLogin.translationZ = 0F
            }
            return false
        }
    }

    private fun saveDrafts() {
        SharedPrefsHelper.save(DRAFT_LOGIN, loginEt.text.toString())
        SharedPrefsHelper.save(DRAFT_USERNAME, usernameEt.text.toString())
    }

    private fun clearDrafts() {
        SharedPrefsHelper.save(DRAFT_LOGIN, "")
        SharedPrefsHelper.save(DRAFT_USERNAME, "")
    }

    private fun prepareUser() {
        val qbUser = QBUser()
        qbUser.login = loginEt.text.toString().trim { it <= ' ' }
        qbUser.fullName = usernameEt.text.toString().trim { it <= ' ' }
        qbUser.password = USER_DEFAULT_PASSWORD
        signIn(qbUser)
    }

    private fun signIn(user: QBUser) {
        showProgressDialog(R.string.dlg_login)
        ChatHelper.login(user, object : QBEntityCallback<QBUser> {
            override fun onSuccess(userFromRest: QBUser, bundle: Bundle?) {
                if (userFromRest.fullName == user.fullName) {
                    loginToChat(user)
                } else {
                    //Need to set password NULL, because server will update user only with NULL password
                    user.password = null
                    updateUser(user)
                }
            }

            override fun onError(e: QBResponseException) {
                if (e.httpStatusCode == UNAUTHORIZED) {
                    signUp(user)
                } else {
                    hideProgressDialog()
                    showErrorSnackbar(R.string.login_chat_login_error, e, View.OnClickListener { signIn(user) })
                }
            }
        })
    }

    private fun updateUser(user: QBUser) {
        ChatHelper.updateUser(user, object : QBEntityCallback<QBUser> {
            override fun onSuccess(qbUser: QBUser, bundle: Bundle?) {
                loginToChat(user)
            }

            override fun onError(e: QBResponseException) {
                hideProgressDialog()
                showErrorSnackbar(R.string.login_chat_login_error, e, null)
            }
        })
    }

    private fun loginToChat(user: QBUser) {
        //Need to set password, because the server will not register to chat without password
        user.password = USER_DEFAULT_PASSWORD
        ChatHelper.loginToChat(user, object : QBEntityCallback<Void> {
            override fun onSuccess(void: Void?, bundle: Bundle?) {
                SharedPrefsHelper.saveQbUser(user)
                if (!chbSave.isChecked) {
                    clearDrafts()
                }
                DialogsActivity.start(this@LoginActivity)
                finish()
                hideProgressDialog()
            }

            override fun onError(e: QBResponseException) {
                hideProgressDialog()
                showErrorSnackbar(R.string.login_chat_login_error, e, null)
            }
        })
    }

    private fun signUp(user: QBUser) {
        SharedPrefsHelper.removeQbUser()
        QBUsers.signUp(user).performAsync(object : QBEntityCallback<QBUser> {
            override fun onSuccess(p0: QBUser?, p1: Bundle?) {
                hideProgressDialog()
                signIn(user)
            }

            override fun onError(exception: QBResponseException?) {
                hideProgressDialog()
                showErrorSnackbar(R.string.login_sign_up_error, exception, null)
            }
        })
    }

    private inner class TextWatcherListener(private var editText: EditText) : TextWatcher {

        private var timer = Timer()

        override fun beforeTextChanged(charSequence: CharSequence?, start: Int, count: Int, after: Int) {

        }

        override fun onTextChanged(charSequence: CharSequence?, start: Int, before: Int, count: Int) {
            val text = charSequence.toString().replace("  ", " ")
            if (editText.text.toString() != text) {
                editText.setText(text)
                editText.setSelection(text.length)
            }
            validateFields()
        }

        override fun afterTextChanged(s: Editable?) {
            timer.cancel()
            timer = Timer()
            timer.schedule(object : TimerTask() {
                override fun run() {
                    saveDrafts()
                }
            }, 300)
        }
    }
}